# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [3]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # TODO: Implement Function
    source_lang = source_text.split('\n')
    target_lang = target_text.split('\n')
    source_id_text = [
        [source_vocab_to_int[word] for word in line.split()] 
        for line in source_lang]
    target_id_text = [
        [target_vocab_to_int[word] for word in line.split()] + 
        [target_vocab_to_int['<EOS>']] 
        for line in target_lang]
    return source_id_text, target_id_text

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper
import problem_unittests as tests

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.1.0


/home/osboxes/anaconda3/envs/py36/lib/python3.6/site-packages/ipykernel_launcher.py:15: UserWarning: No GPU found. Please use a GPU to train your neural network.
  from ipykernel import kernelapp as app


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [7]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    inputs = tf.placeholder(tf.int32, [None, None], name = 'input')
    targets = tf.placeholder(tf.int32, [None,None], name = 'targets')
    learning_rate = tf.placeholder(tf.float32, None, name = 'learning_rate')
    keep_prob = tf.placeholder(tf.float32, None, name = 'keep_prob')
    target_sequence_length = tf.placeholder(tf.int32, [None], name = 'target_sequence_length')
    max_target_sequence_length = tf.reduce_max(target_sequence_length, name = 'max_target_sequence_length')
    source_sequence_length = tf.placeholder(tf.int32, [None], name = 'source_sequence_length')
    return inputs, targets, learning_rate, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length 


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [8]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # TODO: Implement Function
    ending = tf.strided_slice(target_data, [0,0], [batch_size, -1], [1,1])
    return tf.concat([tf.fill([batch_size,1], target_vocab_to_int['<GO>']), ending], 1)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [9]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    # TODO: Implement Function
    encode_embed_input = tf.contrib.layers.embed_sequence(rnn_inputs,source_vocab_size,encoding_embedding_size)
    def stacked(rnn_size):
        encode_cell = tf.contrib.rnn.LSTMCell(rnn_size,initializer=tf.random_uniform_initializer(-0.1,0.1,seed=5))
        return encode_cell
    encode_cell = tf.contrib.rnn.MultiRNNCell([stacked(rnn_size) for _ in range(num_layers)])
    encode_output, encode_state = tf.nn.dynamic_rnn(encode_cell, encode_embed_input, sequence_length=source_sequence_length, dtype=tf.float32)
    return encode_output, encode_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [10]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    # TODO: Implement Function
    train_helper = tf.contrib.seq2seq.TrainingHelper(dec_embed_input, target_sequence_length)
    train_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, train_helper, encoder_state, output_layer)
    train_decoder_output, _ = tf.contrib.seq2seq.dynamic_decode(train_decoder,impute_finished=True,maximum_iterations=max_summary_length)
    return train_decoder_output
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


# Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [12]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    # TODO: Implement Function
    start_infer = tf.tile(tf.constant([start_of_sequence_id], dtype=tf.int32), [batch_size])
    helper_infer = tf.contrib.seq2seq.GreedyEmbeddingHelper(dec_embeddings, start_infer, end_of_sequence_id)   
    decoder_infer = tf.contrib.seq2seq.BasicDecoder(dec_cell, helper_infer, encoder_state, output_layer)
    output_infer, _ = tf.contrib.seq2seq.dynamic_decode(decoder_infer, impute_finished=True, maximum_iterations=max_target_sequence_length)
    return output_infer
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [15]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    embedding_sequences = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    decoding_inputs = tf.nn.embedding_lookup(embedding_sequences, dec_input)
    lstm_stack_cell = []
    for _ in range(num_layers):
        basic_lstm_cell = tf.contrib.rnn.BasicLSTMCell(rnn_size)
        lstm_stack_cell.append(basic_lstm_cell) 
    decoding_cell = tf.contrib.rnn.MultiRNNCell(lstm_stack_cell)
    output_layer = Dense(target_vocab_size, kernel_initializer=tf.truncated_normal_initializer(mean=0.0, stddev=0.1))
    with tf.variable_scope('decoding'):
        training_logits = decoding_layer_train(encoder_state, decoding_cell, decoding_inputs, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)
    sequence_initial = target_vocab_to_int['<GO>']
    sequence_end = target_vocab_to_int['<EOS>']
    with tf.variable_scope('decoding', reuse=True):
        inference_logits = decoding_layer_infer(encoder_state, decoding_cell, embedding_sequences, sequence_initial, sequence_end, max_target_sequence_length, target_vocab_size, output_layer, batch_size, keep_prob)
    return training_logits, inference_logits
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [20]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    encoding_input, encoding_state = encoding_layer(input_data, rnn_size, num_layers, keep_prob, source_sequence_length, source_vocab_size, enc_embedding_size)
    processing_data = process_decoder_input(target_data, target_vocab_to_int, batch_size)
    return decoding_layer(processing_data, encoding_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [35]:
# Number of Epochs
epochs = 2
# Batch Size
batch_size = 560
# RNN Size
rnn_size = 256
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 210
decoding_embedding_size = 210
# Learning Rate
learning_rate = 0.01
# Dropout Keep Probability
keep_probability = 0.7
display_step = 1

### Build the Graph
Build the graph using the neural network you implemented.

In [36]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [37]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [38]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch    1/246 - Train Accuracy: 0.0384, Validation Accuracy: 0.0346, Loss: 4.8957
Epoch   0 Batch    2/246 - Train Accuracy: 0.3109, Validation Accuracy: 0.3433, Loss: 4.5745
Epoch   0 Batch    3/246 - Train Accuracy: 0.2777, Validation Accuracy: 0.3408, Loss: 3.9493
Epoch   0 Batch    4/246 - Train Accuracy: 0.2721, Validation Accuracy: 0.3468, Loss: 3.7819
Epoch   0 Batch    5/246 - Train Accuracy: 0.3167, Validation Accuracy: 0.3459, Loss: 3.5914
Epoch   0 Batch    6/246 - Train Accuracy: 0.3116, Validation Accuracy: 0.3440, Loss: 3.4724
Epoch   0 Batch    7/246 - Train Accuracy: 0.2779, Validation Accuracy: 0.3459, Loss: 3.4602
Epoch   0 Batch    8/246 - Train Accuracy: 0.3060, Validation Accuracy: 0.3459, Loss: 3.3200
Epoch   0 Batch    9/246 - Train Accuracy: 0.2765, Validation Accuracy: 0.3468, Loss: 3.3309
Epoch   0 Batch   10/246 - Train Accuracy: 0.3143, Validation Accuracy: 0.3479, Loss: 3.1908
Epoch   0 Batch   11/246 - Train Accuracy: 0.2882, Validation Accuracy

Epoch   0 Batch   90/246 - Train Accuracy: 0.5229, Validation Accuracy: 0.5580, Loss: 0.8199
Epoch   0 Batch   91/246 - Train Accuracy: 0.5588, Validation Accuracy: 0.5653, Loss: 0.7600
Epoch   0 Batch   92/246 - Train Accuracy: 0.5486, Validation Accuracy: 0.5721, Loss: 0.7969
Epoch   0 Batch   93/246 - Train Accuracy: 0.5678, Validation Accuracy: 0.5748, Loss: 0.7454
Epoch   0 Batch   94/246 - Train Accuracy: 0.5519, Validation Accuracy: 0.5683, Loss: 0.7425
Epoch   0 Batch   95/246 - Train Accuracy: 0.5549, Validation Accuracy: 0.5757, Loss: 0.7356
Epoch   0 Batch   96/246 - Train Accuracy: 0.5596, Validation Accuracy: 0.5858, Loss: 0.7463
Epoch   0 Batch   97/246 - Train Accuracy: 0.5608, Validation Accuracy: 0.5830, Loss: 0.7152
Epoch   0 Batch   98/246 - Train Accuracy: 0.5807, Validation Accuracy: 0.5853, Loss: 0.7112
Epoch   0 Batch   99/246 - Train Accuracy: 0.5556, Validation Accuracy: 0.5782, Loss: 0.7161
Epoch   0 Batch  100/246 - Train Accuracy: 0.5590, Validation Accuracy

Epoch   0 Batch  179/246 - Train Accuracy: 0.6877, Validation Accuracy: 0.7042, Loss: 0.4619
Epoch   0 Batch  180/246 - Train Accuracy: 0.6800, Validation Accuracy: 0.7003, Loss: 0.4689
Epoch   0 Batch  181/246 - Train Accuracy: 0.6558, Validation Accuracy: 0.7018, Loss: 0.4668
Epoch   0 Batch  182/246 - Train Accuracy: 0.6707, Validation Accuracy: 0.6933, Loss: 0.4501
Epoch   0 Batch  183/246 - Train Accuracy: 0.6716, Validation Accuracy: 0.6866, Loss: 0.4463
Epoch   0 Batch  184/246 - Train Accuracy: 0.7057, Validation Accuracy: 0.7071, Loss: 0.4382
Epoch   0 Batch  185/246 - Train Accuracy: 0.7054, Validation Accuracy: 0.7099, Loss: 0.4294
Epoch   0 Batch  186/246 - Train Accuracy: 0.6725, Validation Accuracy: 0.7042, Loss: 0.4548
Epoch   0 Batch  187/246 - Train Accuracy: 0.6971, Validation Accuracy: 0.7183, Loss: 0.4226
Epoch   0 Batch  188/246 - Train Accuracy: 0.7135, Validation Accuracy: 0.7139, Loss: 0.4168
Epoch   0 Batch  189/246 - Train Accuracy: 0.7186, Validation Accuracy

Epoch   1 Batch   24/246 - Train Accuracy: 0.8603, Validation Accuracy: 0.8251, Loss: 0.1772
Epoch   1 Batch   25/246 - Train Accuracy: 0.8242, Validation Accuracy: 0.8382, Loss: 0.1968
Epoch   1 Batch   26/246 - Train Accuracy: 0.8329, Validation Accuracy: 0.8392, Loss: 0.2120
Epoch   1 Batch   27/246 - Train Accuracy: 0.8595, Validation Accuracy: 0.8501, Loss: 0.1889
Epoch   1 Batch   28/246 - Train Accuracy: 0.8594, Validation Accuracy: 0.8397, Loss: 0.1823
Epoch   1 Batch   29/246 - Train Accuracy: 0.8372, Validation Accuracy: 0.8323, Loss: 0.1817
Epoch   1 Batch   30/246 - Train Accuracy: 0.8484, Validation Accuracy: 0.8361, Loss: 0.1748
Epoch   1 Batch   31/246 - Train Accuracy: 0.8332, Validation Accuracy: 0.8335, Loss: 0.1770
Epoch   1 Batch   32/246 - Train Accuracy: 0.8452, Validation Accuracy: 0.8394, Loss: 0.1971
Epoch   1 Batch   33/246 - Train Accuracy: 0.8393, Validation Accuracy: 0.8553, Loss: 0.1742
Epoch   1 Batch   34/246 - Train Accuracy: 0.8429, Validation Accuracy

Epoch   1 Batch  113/246 - Train Accuracy: 0.9085, Validation Accuracy: 0.9099, Loss: 0.0686
Epoch   1 Batch  114/246 - Train Accuracy: 0.9269, Validation Accuracy: 0.9121, Loss: 0.0685
Epoch   1 Batch  115/246 - Train Accuracy: 0.8940, Validation Accuracy: 0.9127, Loss: 0.0764
Epoch   1 Batch  116/246 - Train Accuracy: 0.9180, Validation Accuracy: 0.9140, Loss: 0.0746
Epoch   1 Batch  117/246 - Train Accuracy: 0.9033, Validation Accuracy: 0.9127, Loss: 0.0749
Epoch   1 Batch  118/246 - Train Accuracy: 0.9012, Validation Accuracy: 0.9073, Loss: 0.0737
Epoch   1 Batch  119/246 - Train Accuracy: 0.9038, Validation Accuracy: 0.9151, Loss: 0.0666
Epoch   1 Batch  120/246 - Train Accuracy: 0.8997, Validation Accuracy: 0.9213, Loss: 0.0759
Epoch   1 Batch  121/246 - Train Accuracy: 0.9106, Validation Accuracy: 0.9174, Loss: 0.0728
Epoch   1 Batch  122/246 - Train Accuracy: 0.9111, Validation Accuracy: 0.9178, Loss: 0.0636
Epoch   1 Batch  123/246 - Train Accuracy: 0.9229, Validation Accuracy

Epoch   1 Batch  202/246 - Train Accuracy: 0.9339, Validation Accuracy: 0.9308, Loss: 0.0480
Epoch   1 Batch  203/246 - Train Accuracy: 0.9378, Validation Accuracy: 0.9205, Loss: 0.0406
Epoch   1 Batch  204/246 - Train Accuracy: 0.9411, Validation Accuracy: 0.9316, Loss: 0.0472
Epoch   1 Batch  205/246 - Train Accuracy: 0.9319, Validation Accuracy: 0.9318, Loss: 0.0507
Epoch   1 Batch  206/246 - Train Accuracy: 0.9349, Validation Accuracy: 0.9396, Loss: 0.0480
Epoch   1 Batch  207/246 - Train Accuracy: 0.9247, Validation Accuracy: 0.9330, Loss: 0.0496
Epoch   1 Batch  208/246 - Train Accuracy: 0.9359, Validation Accuracy: 0.9313, Loss: 0.0476
Epoch   1 Batch  209/246 - Train Accuracy: 0.9341, Validation Accuracy: 0.9434, Loss: 0.0446
Epoch   1 Batch  210/246 - Train Accuracy: 0.9448, Validation Accuracy: 0.9343, Loss: 0.0412
Epoch   1 Batch  211/246 - Train Accuracy: 0.9176, Validation Accuracy: 0.9405, Loss: 0.0500
Epoch   1 Batch  212/246 - Train Accuracy: 0.9346, Validation Accuracy

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [39]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [40]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [41]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # TODO: Implement Function
    lower_case_sequence = []
    for word in sentence.lower().split():
        if word in vocab_to_int:
            lower_case_sequence.append(vocab_to_int[word])
        else:
            lower_case_sequence.append(vocab_to_int['<UNK>'])            
    return lower_case_sequence
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [44]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [158, 226, 227, 4, 79, 107, 77]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [180, 281, 65, 237, 194, 82, 214, 146, 1]
  French Words: il a qu'il un vieux camion noir . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.